Re-request a full thinblock on merkleroot check failure

ptschip opened the pull request:
This is a stop gap measure for the time being until a more efficient
refactor can take place.

ptschip committed:
Re-request a full thinblock on merkleroot check failure

This is a stop gap measure for the time being until a more efficient
refactor can take place.

kyuupichan commented on src/thinblock.cpp:
```diff
@@ -378,12 +378,15 @@ bool CXThinBlock::process(CNode *pfrom,
-        return error("Thinblock merkelroot does not match computed merkleroot, peer=%d", pfrom->GetId());
+        vector<CInv> vGetData;
+        vGetData.push_back(CInv(MSG_THINBLOCK, header.GetHash()));
+        pfrom->PushMessage("getdata", vGetData);
```
Please use NetMsgType::GETDATA here (and at the other place in this file).

I approve with that change; and I think this should be considered for backport to release as it solves the banning DoS of #485 until we solve the collision issue.

kyuupichan commented:
I've been running this successfully for about an hour without issues.  Make check passes.  Of course, getting a bad merkleroot block hasn't happened, and might not for months.

kyuupichan commented:
Just happened:

https://hastebin.com/ufevudotoy.pas

The log message could do with a newline

ptschip commented:
@kyuppican

you're right,  i think i'm going refactor this whole thing anyway...as a
stop gap for us to keep testing it works fine but it does break our
re-request functionality.
although not much impact, we only do 1 or 2 rerequests a day anyway, but
still it's not really working the way it was intended.  I'll get on to
it once i finish the 486

On 28/04/2017 9:49 PM, Neil wrote:
>
> Just happened:
>
> https://hastebin.com/ufevudotoy.pas
>
> The log message could do with a newline
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/482#issuecomment-298147076>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0CkrTL-w-IHI0Kwm30IFk8QFoY4_ks5r0sFPgaJpZM4NJtc8>.
>

ptschip commented:
@kyuupichan  I uploaded the refactor...you can run it as is, but it requires manual testing and in the morning I"ll upload the test scenarios I've been doing.   Basically what this refactor does is grabs all the full tx hashes and checks the merkle root before attempting to reconstruct the block.  This way we don't break our re-request functionality and everything works as it did before.

sickpig commented:
@ptschip travis is failing due to `parallel.py` not finishing (looping).

the auto clang format checking finding an inconsistency in `thinblock.cpp`, see:
https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/227503334#L1177

ptschip commented:
@sickpig updated the formatting...the issue with parallel.py is known..it's rare and is a bug in PV which i haven't had time to fix.

ptschip commented:
Added one more commit ...I beleive this PR is complete now and ready for review.

kyuupichan commented on src/thinblock.cpp:
```diff
@@ -224,6 +232,126 @@ CXThinBlockTx::CXThinBlockTx(uint256 blockHash, vector<CTransaction> &vTx)
+                    setPreVerifiedTxHash.insert(hash);
+                else if (inMissingTx)
+                    tx = pfrom->mapMissingTx[hash.GetCheapHash()];
+            }
```
This block of code already exists around line 100 and again around line 360.  They should really be collapsed to one.

kyuupichan commented on src/thinblock.cpp:
```diff
@@ -224,6 +232,126 @@ CXThinBlockTx::CXThinBlockTx(uint256 blockHash, vector<CTransaction> &vTx)
+        {
+            pfrom->thinBlockHashes[i] = val->second.GetHash();
+        }
+        count++;
```
count appears unused?

kyuupichan replied:
What does it mean if it is end?

ptschip commented:
After much discussion with @kyuupichan I'm reverting this to the original single commit.

ptschip committed:
Clear out xthin/thinblock data we no longer need before processing

the block.  Some of this data can be substantial if the memory
pool is large, particularly the mapPartialTxHash.

Commit:
Use global consts for "getdata"

ptschip commented:
just rebasing on 504 to get rid of this test crash error.

kyuupichan commented:
@ptschip I think the rebase lost my change

ptschip commented:
@kyuupichan that's very strange...do you want to rebumit..i'll merge again.

ptschip commented:
oh, i see i didn't pull down your merge first before i rebased , resubmit and i'll merge again.
